### PR TITLE
feat(zones): adds an available services listing tab for zone ingresses

### PR DESCRIPTION
**Pull request by johncowen**
Adds a listing tab for ZoneIngress AvailableServices.

~I've not added any e2e testing for this as yet, I will do once I get feedback whether this is what we want to show here as a first iteration. lmk and I'll write up some e2e tests for it if its good.~ e2e tests added

https://deploy-preview-1607--kuma-gui.netlify.app/gui/zones/driver-0/ingresses/monitor-0/services

Closes https://github.com/kumahq/kuma-gui/issues/1388

### CODE REVIEW

File: src/app/zone-ingresses/views/item/ServicesView.vue [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      <template #title>
+        <h2>
+          <RouteTitle
+            :title="t('zone-ingresses.routes.item.navigation.zone-ingress-services-view')"
+            :render="true"
+          />
+        </h2>
+      </template>
+
+      <KCard class="mt-4">
```
Comment (kleinfreund) [inline src/app/zone-ingresses/views/item/ServicesView.vue:16]: **Suggestion**: This shouldn’t have a margin.

```suggestion
      <KCard>
```

File: src/app/zone-ingresses/views/item/ServicesView.vue [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        <template #body>
+          <template
+            v-for="items in [
+              ...(props.data?.zoneIngress?.availableServices ? [props.data.zoneIngress.availableServices] : [])
```
Comment (kleinfreund) [inline src/app/zone-ingresses/views/item/ServicesView.vue:20]: **Suggestion**: Please use a computed property here. They should be preferred over in-template operations like this. For one, they results are cached if their dependencies don’t change and avoid needless Vue rendering cycles. In-template logic breaks that. 

For another, this is just terrible to read.

File: src/app/zone-ingresses/views/item/ServicesView.vue [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                { label: 'Protocol', key: 'protocol' },
+                { label: 'No. Instances', key: 'instances' },
+                { label: 'Actions', key: 'actions', hideLabel: true },
+              ]"
+              :items="items"
+            >
+              <template #name="{ row: item }: {row: AvailableService}">
+                <template
+                  v-for="service in [
+                    findTag(item.tags, 'kuma.io/service') ?? t('common.collection.none')
```
Comment (kleinfreund) [inline src/app/zone-ingresses/views/item/ServicesView.vue:38]: **Suggestion**: Since we need to process the availableServices data further, we should use a computed property and include the results of `findTag` as part of a new array.

**Review by kleinfreund — CHANGES_REQUESTED** [review]

File: src/app/zone-ingresses/views/item/ServicesView.vue (hunk 1 above)
Comment (johncowen) [reply src/app/zone-ingresses/views/item/ServicesView.vue:16]: Oh! I copied/pasta'ed the wrong one to build from 😆 the Yaml/Config view had this also, I've removed it from there, guessing that one doesn't need it either? Let me know if so and I can put it back (also removed the same from in egresses). Will push this change in a sec

Comment (kleinfreund) [reply src/app/zone-ingresses/views/item/ServicesView.vue]: It’s wrong in these views as far as I can tell:

- src/app/meshes/views/MeshConfigView.vue
- src/app/zone-egresses/views/item/ConfigView.vue
- src/app/zone-ingresses/views/item/ConfigView.vue
- src/app/zones/views/item/ConfigView.vue

Comment (johncowen) [reply src/app/zone-ingresses/views/item/ServicesView.vue]: K I removed those other two too, that pesky ConfigView 😅

File: src/app/zone-ingresses/sources.ts [hunk 4]
```diff
@@ -38,8 +39,11 @@ export const sources = (api: KumaApi) => {
       res.items = res.items.filter((item) => {
         return item.zoneIngress.zone === name
       })
-      res.total = res.items.length
-      return res
+      return {
+        ...res,
+        total: res.items.length,
```
Comment (kleinfreund) [inline src/app/zone-ingresses/sources.ts:44]: **Suggestion**: Can we leave a note here saying that this needs to change once we can filter this API by Zone name? Currently, this is (both in `master` and in this PR) incorrect if there is more than one page of results (total != items.length).

File: src/app/zone-ingresses/views/item/ServicesView.vue [hunk 5]
```diff
@@ -15,102 +15,95 @@
...
-              </template>
-
-              <template #mesh="{ row: item }: {row: AvailableService}">
+            <template #name="{ row: item }: {row: AvailableService}">
+              <template
+                v-for="service in [item.tags?.['kuma.io/service'] ?? t('common.collection.none')]"
+                :key="service"
+              >
                 <RouterLink
+                  v-if="service.length > 0"
```
Comment (kleinfreund) [inline src/app/zone-ingresses/views/item/ServicesView.vue:35]: **Issue**: This can never be `false` unless the service annotation is actually the empty string. If the purpose of this is to be `true` if the service annotation is there, then line 31 above should use `?? null` instead of a `t(...)` result which can’t be predictable.

**Review by kleinfreund — CHANGES_REQUESTED** [review]

File: src/app/zone-ingresses/sources.ts (hunk 4 above)
Comment (johncowen) [reply src/app/zone-ingresses/sources.ts:47]: I added some comment here and in egresses.

Because we are filtering on the frontend. We kinda accepted that for the moment that is ok seeing as ingresses/egresses for a specific zone is never really gonna get past page 1.

File: src/app/zone-ingresses/views/item/ServicesView.vue (hunk 5 above)
Comment (johncowen) [reply src/app/zone-ingresses/views/item/ServicesView.vue]: Pretty sure these conditionals would never happen anyway, guessing that in order for one of these to be an `AvailableService` it would need to had a tag `kuma.io/service`. I'm just checking in with backend now, but I removed all this and altered the type to try to reflect this a little better. I'll wait for confirmation before merging this tho.

Oh I also changed `common.none` to be `''` instead of `' '`

Comment (johncowen) [reply src/app/zone-ingresses/views/item/ServicesView.vue]: Just had confirmation that assuming that `kuma.io/service` is always set on `AvailableService` is ok

File: src/app/zone-ingresses/views/item/ServicesView.vue (hunk 3 above)
Comment (johncowen) [reply src/app/zone-ingresses/views/item/ServicesView.vue]: I removed `findTag` entirely, as I figured we could just access the property here

File: src/app/zone-ingresses/views/item/ServicesView.vue (hunk 2 above)
Comment (johncowen) [reply src/app/zone-ingresses/views/item/ServicesView.vue]: I swapped this out by guaranteeing `availableServices` is always set in the data layer before it ever gets here. Now I just access the property so no need for extra caching or checking for things

**Review by kleinfreund — APPROVED** [review]

**Comment by johncowen** [comment]
I was about to merge but I noticed I'd meant to add some basic e2e tests here so I've done that, I'll leave this hanging for a little while in case but otherwise I'll go ahead and merge

**Comment by lahabana** [comment]
@johncowen how can I get a zone with ingresses? I don't seem to be able to find one...

**Comment by kleinfreund** [comment]
> @johncowen how can I get a zone with ingresses? I don't seem to be able to find one...

I think you’re looking for this information: https://github.com/kumahq/kuma-gui/pull/1586#discussion_r1348567841

TL;DR: Set a cookie `KUMA_ZONE_NAME` to the zone you’re looking at for the mocking stuff to know for which zone to provide mocks.
